helm: reparent on vttablet preStop hook

derekperkins opened the pull request:
Attempt to reparent when vttablet stops. I have a couple questions:

1. Should we try to detect if the current tablet is healthy? If not, we may want to do an `EmergencyReparent` or just let Orchestrator take over
1. The docs say that this hook could possibly get called more than once. Is that something we need to worry about?

cc @enisoc @acharis

hmcgonig commented:
Currently if a stop hook fails, kubernetes will allow the resource to be terminated. 

With this in mind, ideally you would have some retry logic in your stop hook so that it will keep retrying until it actually reparents (or maybe some other form of error handling). For example, on reparent failure you could `sleep infinity` and curl out to some form of alerting so an admin can intervene manually, etc.

There have also been a few scenarios where we've seen reparents cause vttablet to crash, leaving the cluster without a master (luckily orchestrator saves us in this case). 

I wrote our internal stop hook logic and had to think about all this stuff at one point, so I'm very interested to get other's thoughts on the right course of action for both the above cases.

dkhenry commented on helm/vitess/templates/_vttablet.tpl:
```diff
@@ -377,6 +377,36 @@ spec:
+                -topo_global_server_address="etcd-global-client.{{ $namespace }}:2379"
+                -logtostderr=true
+                -stderrthreshold=0
+                PlannedReparentShard -keyspace_shard={{ $keyspace.name }}/{{ $shard.name }}
```
do you want to set `-avoid_master=` here

derekperkins replied:
Good point

sougou commented:
This has been lingering. Can we work on moving this forward?

derekperkins commented:
Yeah, I'm tackling this after backups.

derekperkins commented:
> There have also been a few scenarios where we've seen reparents cause vttablet to crash, leaving the cluster without a master (luckily orchestrator saves us in this case).

@hmcgonig Which container do you have the `preStop` hook running in? In this PR, I have it on `vttablet`, but it's feeling like a race condition between this running and the mysql container stopping. Maybe there's a race condition that is causing vttablet to crash in your case?

hmcgonig commented:
We run it in vttablet. And youre definitely right, there is a race there, you'll want to add a stop hook to the mysql container that polls the vttablet `/debug/vars` endpoint until it becomes unreachable. That way you can guarantee process shutdown ordering.

derekperkins commented:
@hmcgonig @acharis @leoxlin I've implemented the changes we discussed on our call and they really work great. 

- Since I'm not running the `mysqlctld` container directly, but instead copying Vitess files into a user specified image, I ended up using the busybox binary to do the blocking html call to `/debug/vars`.
- I left a TODO for doing more advanced health checking before reparenting that I'll tackle in a future PR.
- I'm currently only deleting the tablet record in the `preStop` hook if it is the master. Should I also delete the record on non-masters?

I cleaned up and rebased my commits, so if you think it looks good, I think it's ready to get merged. Thanks so much for your help and suggestions! It would have taken me a lot longer to figure out how to handle the race condition, and I probably wouldn't have gotten to deleting the tablet record.

hmcgonig commented on docker/k8s/mysqlctld/Dockerfile:
```diff
@@ -18,3 +18,10 @@ COPY --from=k8s /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificate
 # copy vitess config
 COPY --from=k8s /vt/config /vt/config
+
+RUN apt-get update && \
```
I would `RUN apt-get` before copying configs. Because of how docker layering works, youll need to run this apt-get every time you change config files. Granted, this is a super minor thing and only really affects build times.

hmcgonig commented on docker/k8s/vtctlclient/Dockerfile:
```diff
@@ -1,10 +1,14 @@
 FROM debian:stretch-slim
-COPY --from=base /vt/bin/vtctlclient /usr/bin/
+
+COPY --from=k8s /vt/bin/vtctlclient /usr/bin/
```
Same as above.

hmcgonig commented on docker/k8s/vttablet/Dockerfile:
```diff
@@ -11,14 +11,15 @@ RUN mkdir -p /vt/bin
 
 # Copy binaries
 COPY --from=k8s /vt/bin/vttablet /vt/bin/
+COPY --from=k8s /vt/bin/vtctlclient /vt/bin/
```
Both of these could move too.

derekperkins commented:
@hmcgonig Those are good build suggestions for the Dockerfiles. I haven't worried too much about build speed yet since they're being auto-generated on Docker Hub. I added that to this PR.

@sougou this should be ready to merge now.
